ts-client/src/curve/stable-swap.ts

```diff
@@ -1,70 +1,385 @@
+    if (depegType["marinade"]) {
+      const account = this.extraAccounts.get(
+        EXTRA_ACCOUNTS.marinade[0].toBase58()
+      );
+      const coder = new BorshCoder(MarinadeIDL as any as Idl);
+      const stake = coder.accounts.decode("State", account!.data);
+      const msolPrice = stake.msolPrice as BN;
+      return msolPrice.mul(PRECISION).div(new BN(0x1_0000_0000));
+    }
+    if (depegType["lido"]) {
+      const account = this.extraAccounts.get(
+        EXTRA_ACCOUNTS.solido[0].toBase58()
+      );
+      //https://github.com/mercurial-finance/mercurial-dynamic-amm/blob/main/programs/amm/tests/test_depeg_price.rs#L33
+      const stSolSupply = new BN(account!.data.readBigInt64LE(73).toString());
+      const stSolBalance = new BN(account!.data.readBigInt64LE(81).toString());
+      return stSolBalance.mul(PRECISION).div(stSolSupply);
+    }
+    throw new Error("UnsupportedBasePool");
+  }
+
+  private updateDepegInfoIfExpired() {
+    if (!this.depeg.depegType["none"]) {
+      const expired =
+        this.onChainTime >
+        this.depeg.baseCacheUpdated.add(BASE_CACHE_EXPIRE).toNumber();
+      if (expired) {
+        this.depeg.baseVirtualPrice = this.getBasePoolVirtualPrice(
+          this.depeg.depegType
+        );
+        this.depeg.baseCacheUpdated = new BN(this.onChainTime);
+      }
+    }
+  }
+
+  private pegTokenB(tokenBAmount: BN): BN {
+    if (!this.depeg.depegType["none"]) {
+      return tokenBAmount.mul(this.depeg.baseVirtualPrice);
+    }
+    return tokenBAmount;
+  }
+
+  private pegTokenA(tokenAAmount: BN): BN {
+    if (!this.depeg.depegType["none"]) {
+      return tokenAAmount.mul(PRECISION);
+    }
+    return tokenAAmount;
+  }
+
+  private revertPegTokenB(tokenBAmount: BN): BN {
+    if (!this.depeg.depegType["none"]) {
+      return tokenBAmount.div(this.depeg.baseVirtualPrice);
+    }
+    return tokenBAmount;
+  }
+
+  private revertPegTokenA(tokenAAmount: BN): BN {
+    if (!this.depeg.depegType["none"]) {
+      return tokenAAmount.div(PRECISION);
+    }
+    return tokenAAmount;
+  }
+
+  private normalizeTokenA(tokenAAmount: BN): BN {
+    const { tokenAMultiplier } = this.tokenMultiplier;
+    return tokenAAmount.mul(tokenAMultiplier);
+  }
+
+  private normalizeTokenB(tokenBAmount: BN): BN {
+    const { tokenBMultiplier } = this.tokenMultiplier;
+    return tokenBAmount.mul(tokenBMultiplier);
+  }
+
+  private denormalizeTokenA(tokenAAmount: BN): BN {
+    const { tokenAMultiplier } = this.tokenMultiplier;
+    return tokenAAmount.div(tokenAMultiplier);
+  }
+
+  private denormalizeTokenB(tokenBAmount: BN): BN {
+    const { tokenBMultiplier } = this.tokenMultiplier;
+    return tokenBAmount.div(tokenBMultiplier);
+  }
+
+  private upscaleTokenA(tokenAAmount: BN): BN {
+    const normalizedTokenAAmount = this.normalizeTokenA(tokenAAmount);
+    return this.pegTokenA(normalizedTokenAAmount);
+  }
+
+  private downscaleTokenA(tokenAAmount: BN): BN {
```

Comment:
I think we can wrap all logic: for normalize + pegPrice in 1 function

Reply:
Fixed